docs/design/tracing.md

```diff
@@ -0,0 +1,148 @@
+# Tracing Design
+
+* **Type**: Design
+* **Author(s)**: Ian Botsford
+
+# Abstract
+
+Tracing describes the emission of logging and metric events in a structured manner for the purposes of analyzing SDK
+performance and debugging. This document presents a design for how tracing will work in the SDK.
+
+# Concepts
+
+The following terms are defined:
+
+* **Trace span**: A logical grouping of tracing events that encompasses some operation. Trace spans may be subdivided
+  into child spans which group a narrower set of events within the context of the parent. Trace spans are hierarchical;
+  events that occur within one span also logically occur within the ancestors of that span.
+
+* **Trace probe**: A receiver for tracing events. A probe is notified when new events occur within a span and may take
+  appropriate action to route the event (e.g., forward to a downstream logging/metrics framework, print to the console,
+  write to a file, etc.).
+
+# Design
+
+The following components provide tracing support:
+
+## Tracer
+
+A `Tracer` is a top-level provider of tracing capabilities. It bridges trace spans (into which events are emitted) and
+trace probes (which receive events and handle them accordingly). Typically, each service client will have its own
+internal `Tracer` instance. That `Tracer` need not be publicly accessible but must be configurable with a trace probe
+and client name at service client construction.
+
+The `Tracer` interface is specified as:
+
+```kotlin
+interface Tracer {
+    fun createRootSpan(id: String): TraceSpan
+}
+```
+
+A `Tracer` provides root spans for a service client, into which all events over the lifetime of an operation will be
+emitted. Child spans can be created as mentioned below in the [Trace Span](#trace-span) section.
+
+**Note**: The interface does not specify how trace probes will be configured or utilized. These are implementation
+details of the tracer and aren't necessary in the public interface.
+
+## Trace Span
+
+A `TraceSpan` is a logical grouping of tracing events that are associated with some operation. Spans may be subdivided
+into child spans which group a narrower set of events within the context of the parent. Spans are hierarchical; events
+that occur within one span also logically occur within the ancestors of that span.
+
+The `TraceSpan` interface is specified as:
+
+```kotlin
+interface TraceSpan : Closeable {
+    val id: String
+    val parent: TraceSpan?
+
+    fun child(id: String): TraceSpan
+    fun postEvents(events: Iterable<TraceEvent>)
```

Comment:
comment: I know we've discussed this before but I was reviewing the tracing section of the ref architecture work and it includes both `emitEvent(event: TraceEvent)` and `emitEvents(events: Iterable<TraceEvent>)`. I'd argue we should remove the plural version and only accept the singular `postEvent(event: TraceEvent)`. This is based on (1) we have yet to have an actual need for the plural when emitting events; I don't expect this will change. (2) The reason for taking plural as explained in the ref arch is performance. That's fine _but_ it shouldn't really be a concern of a `TraceSpan`. It is more appropriate for trace probes to deal with batching/queuing of events (if necessary). I'm rather convinced now that it doesn't belong in this API (or at least we should support both) and I will be recommending the update to the architecture doc as well.

Reply:
Discussed offline and came to the conclusion that singular APIs will be better for the average case than plural (and cleaner than having both). Will change!